include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -17,131 +20,449 @@ namespace hlsl
     {
-        const T NdotTorR = r ? computeNdotT(backside, NdotI2, rcpOrientedEta2) : NdotI;
-        return doReflectRefract(r, I, N, NdotI, NdotTorR, rcpOrientedEta);
+        return N * (NdotI * rcpOrientedEta + NdotT) - rcpOrientedEta * I;
     }
 
     vector_type I;
     vector_type N;
-    bool backside;
-    scalar_type NdotI;
-    scalar_type NdotI2;
-    scalar_type rcpOrientedEta;
-    scalar_type rcpOrientedEta2;
 };
 
+template<typename T NBL_PRIMARY_REQUIRES(concepts::Scalar<T>)
+struct ReflectRefract
+{
+    using this_t = ReflectRefract<T>;
+    using vector_type = vector<T, 3>;
+    using scalar_type = T;
+
+    // when you know you'll reflect
+    scalar_type getNdotR() NBL_CONST_MEMBER_FUNC
+    {
+        return refract.getNdotI();
+    }
+
+    // when you know you'll refract
+    scalar_type getNdotT(const scalar_type rcpOrientedEta) NBL_CONST_MEMBER_FUNC
+    {
+        return refract.getNdotT(rcpOrientedEta*rcpOrientedEta);
+    }
+
+    scalar_type getNdotTorR(const bool doRefract, const scalar_type rcpOrientedEta) NBL_CONST_MEMBER_FUNC
+    {
+        return hlsl::mix(getNdotR(), getNdotT(rcpOrientedEta), doRefract);
+    }
+
+    vector_type operator()(const bool doRefract, const scalar_type rcpOrientedEta, NBL_REF_ARG(scalar_type) out_IdotTorR) NBL_CONST_MEMBER_FUNC
+    {
+        scalar_type NdotI = getNdotR();
+        const scalar_type a = hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract);
+        const scalar_type b = NdotI * a + getNdotTorR(doRefract, rcpOrientedEta);
+        // assuming `I` is normalized
+        out_IdotTorR = NdotI * b - a;
+        return refract.N * b - refract.I * a;
+    }
+
+    vector_type operator()(const bool doRefract, const scalar_type rcpOrientedEta) NBL_CONST_MEMBER_FUNC
+    {
+        scalar_type dummy;
+        return operator()(doRefract, rcpOrientedEta, dummy);
+    }
+
+    vector_type operator()(const scalar_type NdotTorR, const scalar_type rcpOrientedEta) NBL_CONST_MEMBER_FUNC
+    {
+        scalar_type NdotI = getNdotR();
+        bool doRefract = ComputeMicrofacetNormal<scalar_type>::isTransmissionPath(NdotI, NdotTorR);
+        return refract.N * (NdotI * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract)) + hlsl::mix(NdotI, NdotTorR, doRefract)) - refract.I * (hlsl::mix<scalar_type>(1.0f, rcpOrientedEta, doRefract));
+    }
+
+    Refract<scalar_type> refract;
+};
+
+
+namespace fresnel
+{
+
+#define NBL_CONCEPT_NAME Fresnel
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (fresnel, T)
+NBL_CONCEPT_BEGIN(1)
+#define fresnel NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_TYPE)(T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::vector_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((fresnel()), ::nbl::hlsl::is_same_v, typename T::vector_type))
```

Comment:
the signature of the fresnel should be `operator()(const scalar_type clampedCosTheta)` the cosines should not be stored in the fresnel!